fixes "A Boy's Dream" blocking if a player has the odonto ra/ex

SirGouki opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

Fixes #2729 by checking for the appropriate CS value during the trade instead of checking if they don't have the odonto for the giant shell bug trade.

## Steps to test these changes

Start "A Boy's Dream" quest for PLD AF.
Get some giant shell bugs but don't trade one to Ailbeche yet.
Go Get the odonto ra/ex as per usual for quest.

Before: If you did this before, Ailbeche would not progress the quest due to the lua checking that the player doesn't have the Odonto ra/ex needed for this quest.

After: Ailbeche will not accept the giant shell bug and properly progress the quest, talking to him will repeat the previous message he gives after a trade (In testing, he'd say "I wish  could go fishing with my father..." again before repeating his last appropriate statement, so players may have to talk to him more than once if they need him to repeat it for some reason).  He will not accept a repeat trade of the giant shell bug / odonto.

This same fix also prevents players accidently rolling back quest progression by trading the shell bug to Ailbeche after the odonto.

SirGouki committed:
fixes #2729

SirGouki committed:
fixed typo

SirGouki commented:
How I tested:
Did the previous AF quest before this.
Gave myself a giant shell bug and odonto via !additem <id> 1
Progressed the quest the way a player would normally, except using !gotoid to the relevant npcs involved.

This was tested on a clean LAN LSB server

zach2good commented:
Can this have a more descriptive name please?

SirGouki commented:
is that better?

dallano commented on scripts/zones/Northern_San_dOria/npcs/Ailbeche.lua:
```diff
@@ -15,9 +15,9 @@ local entity = {}
 entity.onTrade = function(player, npc, trade)
     if (player:getCharVar("aBoysDreamCS") >= 3) then
-        if (trade:hasItemQty(17001, 1) == true and trade:getItemCount() == 1 and player:hasItem(4562) == false) then
+        if (trade:hasItemQty(17001, 1) == true and trade:getItemCount() == 1 and player:getCharVar("aBoysDreamCS") == 3) then
```
We can now reduce the first two if statement checks into:
```suggestion
        if npcUtil.tradeHasExactly(trade, xi.items.GIANT_SHELL_BUG) and player:getCharVar("aBoysDreamCS") == 3) then
```
You'll want to make sure this lua has the following library:
require("scripts/globals/npc_util")
You may have to add GIANT_SHELL_BUG to items.lua. Very easy to enter in, but absolutely necessary.
Same goes for below, but the item would be xi.items.ODONTOTYRANNUS

SirGouki replied:
I just modified what I knew.  I'll check this out in a bit.

dallano replied:
Look great! Thank you :) it's always good to fix these kinds things when we come by them.

SirGouki committed:
Adds ODONTOTYRANNUS and GIANT_SHELL_BUG to items.lua
changes Ailbeche and Zaldon to use xi.items.ODONTOTYRANNUS/GIANT_SHELL_BUG, where appropriate,
and npcUtils.tradeHasExactly() to reduce if checks

dallano commented on scripts/zones/Northern_San_dOria/npcs/Ailbeche.lua:
```diff
@@ -15,9 +15,9 @@ local entity = {}
 
```
This lua needs the following library:
require("scripts/globals/npc_util")

SirGouki replied:
WILCO

Edit:
I had it in my local version when I tested and forgot to add it to the branch.

SirGouki committed:
fixed missing require

TeoTwawki commented:
> is that better?
> 
> sorry it pulled the name from my commit message.

normally, both the commit message AND the pr title should be like a summary of what you've actually done (commit messages are suggested to be like you are giving git orders `cook fries before putting them in the basket`) and then "`fixes #number` would be tacked on the next line of the commit message and in the description body of the pr for best results. We don't enforce a particular style, but we do ask for descriptive naming in both PR titles and in commit messages so we can look in the history of both local copies and github merges and see what it actually was at a glance.


Links that may help with writing meaningful messages, and a short sample from them:

https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/


> To come up with thoughtful commits, consider the following:
> 
> - Why have I made these changes?
> - What effect have my changes made?
> - Why was the change needed?
> - What are the changes in reference to?
> 
> Assume the reader does not understand what the commit is addressing. They may not have access to the story addressing the detailed background of the change.

https://initialcommit.com/blog/git-commit-messages-best-practices

> As a general rule, your messages should start with a single line that’s no more than about 50 characters and that describes the changeset concisely, followed by a blank line, followed by a more detailed explanation.

And finally, this is from the author of git himself:

https://github.com/torvalds/subsurface-for-dirk/blob/a48494d2fbed58c751e9b7e8fbff88582f9b2d02/README#L88

> Header line: explain the commit in one line (use the imperative)
> 
> Body of commit message is a few lines of text, explaining things
> in more detail, possibly giving some background about the issue
> being fixed, etc etc.
> 
> The body of the commit message can be several paragraphs, and
> please do proper word-wrap and keep columns shorter than about
> 74 characters or so. That way "git log" will show things
> nicely even when it's indented.
> 
> Make sure you explain your solution and why you're doing what you're
> doing, as opposed to describing what you're doing. Reviewers and your
> future self can read the patch, but might not understand why a
> particular solution was implemented.

SirGouki commented:
@TeoTwawki ok I'll try to be more mindful of this.

`@anyone`
Is this ready to be merged yet?  I'm waiting on this to update my stuff so I can move on to other things

claywar commented:
You've changed two conditions here requiring a charvar that I don't see a counterpart to update, which means in order to review we'll need to go through the entire line to verify:
![image](https://user-images.githubusercontent.com/81713309/195459657-fcf0beee-09b8-41d1-bda4-625acc8a8890.png)

Also, there is a conflict that needs to be resolved, so a rebase will be necessary.

SirGouki commented:
> You've changed two conditions here requiring a charvar that I don't see a counterpart to update, which means in order to review we'll need to go through the entire line to verify: ![image](https://user-images.githubusercontent.com/81713309/195459657-fcf0beee-09b8-41d1-bda4-625acc8a8890.png)
> 
> Also, there is a conflict that needs to be resolved, so a rebase will be necessary.

I don't quite get the first part of this, I PRed this because of the comments in the relevant issue.

As for the conflict, this is due to how long it takes to get reviews when everyone is busy... I can drop the PR and redo it when we can decide what needs to be done, but I'd prefer live discussion with someone on Discord over waiting for reviews, as then I can just PR something that'll get merged instead of risking conflicts because its taking so long.  Been told thats annoying though...

zach2good commented:
It isn't merge yet because I didn't finish all possible configurations before I went on holiday, but I've started a git guide with steps for how to rebase and squash:

https://github.com/LandSandBoat/lsb-wiki/blob/560d0eb067c31789f1c7c76629dfc57558eda8fa/Git-Guide.md

zach2good commented:
Seems like there's a button to work through the merge conflict online.

![Screenshot_20221013-074538](https://user-images.githubusercontent.com/1389729/195503905-62d65ae1-4ba6-46c2-901e-fa470fa8b50c.png)

As with all tech and programming things, there are seas of good tutorials on YouTube

Xaver-DaRed committed:
Merge branch 'base' into ABoysDream

Xaver-DaRed reviewed (approved):
Conflict resolved.

Looks good to me.

zach2good reviewed (approved):
Seems sane, remember to squash on merge, whoever merges

SirGouki commented:
> It isn't merge yet because I didn't finish all possible configurations before I went on holiday, but I've started a git guide with steps for how to rebase and squash:
> 
> https://github.com/LandSandBoat/lsb-wiki/blob/560d0eb067c31789f1c7c76629dfc57558eda8fa/Git-Guide.md

I thought it was the reviews holding it up cause it was still saying it needed review.
